# Exploratory Data Analysis of Craigslist Cars Listing

This notebook explores a dataset of cars listing from Craigslist, which was originally published on Kaggle by Austin Reese. 
The dataset contains information about over 400k listings, including details such as the vehicle's make, model, year, price, location, and description.

The dataset was collected using the Craigslist API and scraped from Craigslist's website. 
The project was taken offline around December 2020 but the posting date of listings seems to range from April 2021 to May 2021 (TBC)

The dataset is available for download on Kaggle at the following link: [Craigslist Car and Truck Listings](https://www.kaggle.com/datasets/austinreese/craigslist-carstrucks-data)

The dataset is in CSV format, with one row per listing and one column per variable.

The goal of this notebook is to explore the dataset. We will use data visualization and statistical analysis to answer questions such as:

- What are the most popular vehicle makes and models on Craigslist?
- etc.

<!-- variable TBC
- `id`: a unique identifier for each listing
- `url`: the URL of the listing on Craigslist
- `region`: the region where the listing was posted (e.g. "atlanta", "boston", etc.)
- `region_url`: the URL of the region on Craigslist
- `price`: the price of the vehicle, in USD
- `year`: the year of the vehicle
- `manufacturer`: the manufacturer of the vehicle (e.g. "ford", "honda", etc.)
- `model`: the model of the vehicle (e.g. "focus", "civic", etc.)
- `condition`: the condition of the vehicle (e.g. "new", "like new", "excellent", etc.)
- `cylinders`: the number of cylinders in the vehicle's engine (e.g. "4 cylinders", "6 cylinders", etc.)
- `fuel`: the type of fuel used by the vehicle (e.g. "gas", "diesel", etc.)
- `odometer`: the mileage of the vehicle, in miles
- `title_status`: the title status of the vehicle (e.g. "clean", "salvage", etc.)
- `transmission`: the type of transmission used by the vehicle (e.g. "automatic", "manual", etc.)
- `VIN`: the vehicle identification number (VIN) of the vehicle
- `drive`: the type of drive used by the vehicle (e.g. "4wd", "fwd", etc.)
- `size`: the size of the vehicle (e.g. "full-size", "mid-size", etc.)
- `type`: the type of vehicle (e.g. "sedan", "SUV", etc.)
- `paint_color`: the paint color of the vehicle (e.g. "white", "black", etc.)
- `image_url`: the URL of an image of the vehicle
- `description`: a description of the vehicle, provided by the seller
- `county`: the county where the listing was posted
- `state`: the state where the listing was posted
- `lat`: the latitude of the listing's location
- `long`: the longitude of the listing's location
- `posting_date`: the date when the listing was posted on Craigslist -->

# 0 - Imports

## 0.1 - Libs imports

In [1]:
import pandas as pd

## 0.2 - Data import

In [2]:
df = pd.read_csv('../data/vehicles.csv', encoding="utf-8")

# 1 - Exploration

## 1.1 - Basic info and first look at the data

In [3]:
df.head()

id                                                url  \
0  7222695916  https://prescott.craigslist.org/cto/d/prescott...   
1  7218891961  https://fayar.craigslist.org/ctd/d/bentonville...   
2  7221797935  https://keys.craigslist.org/cto/d/summerland-k...   
3  7222270760  https://worcester.craigslist.org/cto/d/west-br...   
4  7210384030  https://greensboro.craigslist.org/cto/d/trinit...   

                   region                         region_url  price  year  \
0                prescott    https://prescott.craigslist.org   6000   NaN   
1            fayetteville       https://fayar.craigslist.org  11900   NaN   
2            florida keys        https://keys.craigslist.org  21000   NaN   
3  worcester / central MA   https://worcester.craigslist.org   1500   NaN   
4              greensboro  https://greensboro.craigslist.org   4900   NaN   

  manufacturer model condition cylinders  ... size  type paint_color  \
0          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
1          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
2          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
3          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
4          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   

  image_url description county state lat long posting_date  
0       NaN         NaN    NaN    az NaN  NaN          NaN  
1       NaN         NaN    NaN    ar NaN  NaN          NaN  
2       NaN         NaN    NaN    fl NaN  NaN          NaN  
3       NaN         NaN    NaN    ma NaN  NaN          NaN  
4       NaN         NaN    NaN    nc NaN  NaN          NaN  

[5 rows x 26 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 26 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            426880 non-null  int64  
 1   url           426880 non-null  object 
 2   region        426880 non-null  object 
 3   region_url    426880 non-null  object 
 4   price         426880 non-null  int64  
 5   year          425675 non-null  float64
 6   manufacturer  409234 non-null  object 
 7   model         421603 non-null  object 
 8   condition     252776 non-null  object 
 9   cylinders     249202 non-null  object 
 10  fuel          423867 non-null  object 
 11  odometer      422480 non-null  float64
 12  title_status  418638 non-null  object 
 13  transmission  424324 non-null  object 
 14  VIN           265838 non-null  object 
 15  drive         296313 non-null  object 
 16  size          120519 non-null  object 
 17  type          334022 non-null  object 
 18  pain

> Notes:
> - 426.880 entries!
> - Some data are missing, keep an eye on it
> - "county" is always missing 

In [5]:
df["posting_date"].describe()
#df['time'] = pd.to_datetime(df['time'])??

count                       426812
unique                      381536
top       2021-04-23T22:13:05-0400
freq                            12
Name: posting_date, dtype: object

In [6]:
df['manufacturer'].value_counts().head(10)

manufacturer
ford         70985
chevrolet    55064
toyota       34202
honda        21269
nissan       19067
jeep         19014
ram          18342
gmc          16785
bmw          14699
dodge        13707
Name: count, dtype: int64

> Notes: 
> - Ford is the most popular manufacturer, not very surprising haha

In [7]:
df['model'].value_counts().head(5)

model
f-150             8009
silverado 1500    5140
1500              4211
camry             3135
silverado         3023
Name: count, dtype: int64

> Notes:
> - And of course, the most popular car is the F-150 (from Ford)
> - Sales by year: 896,526 (2019), 787,372 (2020) and 726,003 (2021) [Source](https://www.goodcarbadcar.net/ford-f-series-sales-figures/). Yeah, it's huge haha!

## 1.2 - Deeper look into each variable

Ok, let's explore our data and define each column and separate them in what we want to keep or not: 

**KEEP:** ✔️
- `id`: a unique identifier for each listing 
- `region`: the region where the listing was posted (e.g. "atlanta", "boston", etc.)
- `region_url`: the URL of the region on Craigslist (_potentially useful_)
- `price`: the price of the vehicle, in USD
- `year`: the year of the vehicle
- `manufacturer`: the manufacturer of the vehicle (e.g. "ford", "honda", etc.)
- `model`: the model of the vehicle (e.g. "focus", "civic", etc.)
- `condition`: the condition of the vehicle (e.g. "new", "like new", "excellent", etc.)
- `cylinders`: the number of cylinders in the vehicle's engine (e.g. "4 cylinders", "6 cylinders", etc.)
- `fuel`: the type of fuel used by the vehicle (e.g. "gas", "diesel", etc.)
- `odometer`: the mileage of the vehicle, **in miles**
- `title_status`: the title status of the vehicle (e.g. "clean", "salvage", etc.)
- `transmission`: the type of transmission used by the vehicle (e.g. "automatic", "manual", etc.)
- `drive`: the type of drive used by the vehicle (e.g. "4wd", "fwd", etc.)
- `size`: the size of the vehicle (e.g. "full-size", "mid-size", etc.)
- `type`: the type of vehicle (e.g. "sedan", "SUV", etc.)
- `paint_color`: the paint color of the vehicle (e.g. "white", "black", etc.)
- `description`: a description of the vehicle, provided by the seller
- `state`: the state where the listing was posted
- `lat`: the latitude of the listing's location
- `long`: the longitude of the listing's location
- `posting_date`: the date when the listing was posted on Craigslist

**IGNORE:** ❌
- `url`: the URL of the listing on Craigslist (not used + URL from 2021...)
- `image_url`: the URL of an image of the vehicle (not used + URL from 2021...)
- `VIN`: the vehicle identification number (VIN) of the vehicle (UUID already provided by `id`)
- `county`: the county where the listing was posted (all NaN)


Overall, almost all data are potentially useful. Let's now deal with missing values.


In [8]:
pd.set_option('display.max_columns', None)
df.sample(n=4)

id                                                url  \
209113  7307917555  https://saginaw.craigslist.org/ctd/d/chesaning...   
58370   7305951660  https://slo.craigslist.org/ctd/d/grover-beach-...   
425153  7310344718  https://northernwi.craigslist.org/ctd/d/loyal-...   
415613  7308806310  https://appleton.craigslist.org/cto/d/oshkosh-...   

                         region                         region_url  price  \
209113  saginaw-midland-baycity     https://saginaw.craigslist.org  11995   
58370           san luis obispo         https://slo.craigslist.org   6995   
425153              northern WI  https://northernwi.craigslist.org  25995   
415613     appleton-oshkosh-FDL    https://appleton.craigslist.org   5500   

          year manufacturer                model  condition    cylinders fuel  \
209113  2010.0         ford  explorer sport trac        NaN          NaN  gas   
58370   2003.0         ford               ranger  excellent          NaN  gas   
425153  2015.0    chevrolet     silverado 3500hd  excellent  8 cylinders  gas   
415613  2011.0    chevrolet             traverse  excellent  6 cylinders  gas   

        odometer title_status transmission                VIN drive size type  \
209113  123879.0        clean    automatic                NaN   NaN  NaN  NaN   
58370   170284.0        clean    automatic  1FTYR10U23PB39317   rwd  NaN  NaN   
425153  167117.0        clean    automatic  1GB5KYCG4FZ525756   4wd  NaN  NaN   
415613  247000.0        clean    automatic                NaN   fwd  NaN  NaN   

       paint_color                                          image_url  \
209113         NaN  https://images.craigslist.org/00707_25AtYCGeRP...   
58370        black  https://images.craigslist.org/00H0H_88YMt7GZKE...   
425153         NaN  https://images.craigslist.org/00I0I_jjSsmYJ2vc...   
415613         NaN  https://images.craigslist.org/00808_43NqRfRr7Q...   

                                              description  county state  \
209113  WE HAVE OVER 350 VEHICLES IN STOCK!   View Our...     NaN    mi   
58370   2003 *** FORD RANGER  ***    SEE WHY PEOPLE SA...     NaN    ca   
425153  DUALLY 4X4 FLATBED 6.0 GAS GREAT PRICE ONE OWN...     NaN    wi   
415613  2011 Chevy traverse LT. Higher milage, 247000,...     NaN    wi   

              lat        long              posting_date  
209113  43.182400  -84.112200  2021-04-16T16:23:34-0400  
58370   35.122159 -120.626169  2021-04-12T17:15:49-0700  
425153  44.741365  -90.485795  2021-04-21T10:31:00-0500  
415613  44.030400  -88.607000  2021-04-18T12:07:11-0500

In [9]:
total_rows = df.shape[0]

#Get null numbers for each column
null_counts = df.isnull().sum()

summary_df = pd.DataFrame({'Column': null_counts.index})

summary_df['Percentage of Null'] = ((null_counts.values / total_rows) * 100).round(1)

summary_df = summary_df.sort_values('Percentage of Null', ascending=False)

summary_df['Percentage of Null'] = summary_df['Percentage of Null'].map(lambda x: f'{x}%')

summary_df


Column Percentage of Null
21        county             100.0%
16          size              71.8%
9      cylinders              41.6%
8      condition              40.8%
14           VIN              37.7%
15         drive              30.6%
18   paint_color              30.5%
17          type              21.8%
6   manufacturer               4.1%
12  title_status               1.9%
23           lat               1.5%
24          long               1.5%
7          model               1.2%
11      odometer               1.0%
10          fuel               0.7%
13  transmission               0.6%
5           year               0.3%
0             id               0.0%
3     region_url               0.0%
2         region               0.0%
4          price               0.0%
1            url               0.0%
20   description               0.0%
19     image_url               0.0%
22         state               0.0%
25  posting_date               0.0%

Based on the code above, we can see some data are missing. For some of them, it's between 1/3 and all the lines with missing data:

| Column     | Percentage of Null |
|------------|--------------------|
| county     | 100.0%            |
| size       | 71.8%              |
| cylinders  | 41.6%              |
| condition  | 40.8%              |
| VIN        | 37.7%              |
| drive      | 30.6%              |
| paint_color| 30.5%              |

So we can not simply dropna() and loose potentially half (or all!) the data. For now, just know we can drop the column `county` without any remorse.

In [10]:
if 'county' in df.columns:
    df = df.drop(columns=['county'])
print(f"Number of columns remaining: {df.shape[1]}")

Number of columns remaining: 25


As we have a lot of data, we can also drop the lines with NaN in the subset described below: 

In [22]:
df = df.dropna(subset=['manufacturer', 'title_status', "lat", 'long', 'model', 'odometer', 'fuel', 'transmission', 'year', 'id', 'region_url', 'region', 'price', 'description', 'image_url', 'state', 'posting_date'])
print("By doing so, we are losing", round(((total_rows-df.shape[0]) / total_rows) * 100,2), "% of the data. We still get", df.shape[0], "lines to use.")

By doing so, we are losing 10.17 % of the data. We still get 383487 lines to use.


Also, we will be droping the "ignored" columns mentionned earlier in the dedicated script.

## 1.3 - Are all the data types correct?

From the .info() we got earlier, price/year/odometer/county/posting_date may not be the adequate types.

- Price: can not be set as float on Craiglist, so int64 is adequate
- Year: is set a float when years should be int, the code below is going to correct it
- Odometer: can not be set as float on Craiglist, the code below is going to correct it
- County: is always empty and will be ignored so it's ok to leave it as is 
- Posting_date: is set as "object" which is not wrong but datetime data type would allow us to perform better information retrieval, let's fix that


In [ ]:
#NaN ignored by default
df['year'] = pd.to_numeric(df['year'], errors="coerce").astype('Int64')

df['odometer'] = pd.to_numeric(df['odometer'], errors="coerce").astype('Int64')

#ISO 8601 format
df['posting_date'] = pd.to_datetime(df['posting_date'], errors='coerce')

#I need to run this line twice??

In [ ]:
df['posting_date'].dt.year
#returns float bc NaN but dtype is datetime

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
          ..
426875   NaN
426876   NaN
426877   NaN
426878   NaN
426879   NaN
Name: posting_date, Length: 426880, dtype: float64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 25 columns):
 #   Column        Non-Null Count   Dtype                    
---  ------        --------------   -----                    
 0   id            426880 non-null  int64                    
 1   url           426880 non-null  object                   
 2   region        426880 non-null  object                   
 3   region_url    426880 non-null  object                   
 4   price         426880 non-null  int64                    
 5   year          425675 non-null  Int64                    
 6   manufacturer  409234 non-null  object                   
 7   model         421603 non-null  object                   
 8   condition     252776 non-null  object                   
 9   cylinders     249202 non-null  object                   
 10  fuel          423867 non-null  object                   
 11  odometer      422480 non-null  Int64                    
 12  title_status  41

In [ ]:
df['post_day'] = df['posting_date'].dt.day.astype('Int64')
df['post_month'] = df['posting_date'].dt.month.astype('Int64')
df['post_year'] = df['posting_date'].dt.year.astype('Int64')

## 1.4 - Some sanity checks

In [ ]:
df["model"].sample(10)
#it seems to indicate the model is a free text field, it will complicate the task

298881      pickup 2500 sltquad
363727                     328i
318731        cooper countryman
78585       golf sportwagen tsi
26199     rav4 hybrid xse sport
279363                   450 sl
117011                      cts
16165                      edge
378793                   sentra
4267                       f150
Name: model, dtype: object

In [ ]:
print(*set(df["manufacturer"]), sep=" / ")
#it seems to indicate the manufacturer is taken from a list, nice

ford / jaguar / rover / jeep / saturn / mercury / alfa-romeo / land rover / mazda / volkswagen / pontiac / chrysler / hyundai / fiat / gmc / mini / audi / aston-martin / lexus / volvo / acura / lincoln / cadillac / mercedes-benz / mitsubishi / chevrolet / kia / morgan / porsche / ram / dodge / bmw / tesla / nan / harley-davidson / subaru / honda / ferrari / toyota / buick / infiniti / nissan / datsun


In [ ]:
pd.options.display.max_colwidth = 500
print(df.description.sample(10).to_string(index=True))
#a lot of content, yup

30481     Vehicle DescriptionEverything youre looking for and more, our 2017 Kia Forte LX Sedan is sleek and sporty in Silky Silver! Motivated by a robust 2.0 Liter 4 Cylinder that provides 147hp with its responsive 6 Speed Automatic transmission for passing ease. This Front Wheel Drive Sedan accelerates quickly and delivers a sporty feel while also rewarding your wallet with near 38mpg on the highway! Our Forte has a style for miles with a stunning silhouette that is further complemented by beautiful...
334500    2011 BMW X5 xDrive35i     Offered by: Car N Drive - INCREDIBLE DEALS FOR TAX SEASON! — (609) 631-3522 — $9,799       Car N Drive - INCREDIBLE DEALS FOR TAX SEASON!    Year: 2011 Make: BMW Model: X5 Series: xDrive35i VIN: 5UXZV4C56BL415511 Stock #: 415511 Condition: Used Mileage: 139,240  Exterior: Platinum Gray Metallic Interior: Beige Leatherette i Body: Sport Utility 4-dr Transmission: Automatic Engine: 3.0L L6 DOHC 24V      ⭐️ CAR N DRIVE - PAYMENTS AS LOW AS $49/WEEK! ⭐️ 

In [ ]:
print(*set(df["year"]), sep=" / ")

1900 / 1901 / 1902 / 1903 / 1905 / 1909 / 1910 / 1913 / 1915 / 1916 / 1918 / 1920 / 1921 / 1922 / 1923 / 1924 / 1925 / 1926 / 1927 / 1928 / 1929 / 1930 / 1931 / 1932 / 1933 / 1934 / 1935 / 1936 / 1937 / 1938 / 1939 / 1940 / 1941 / 1942 / 1943 / 1944 / 1945 / 1946 / 1947 / 1948 / 1949 / 1950 / 1951 / 1952 / 1953 / 1954 / 1955 / 1956 / 1957 / 1958 / 1959 / 1960 / 1961 / 1962 / 1963 / 1964 / 1965 / 1966 / 1967 / 1968 / 1969 / 1970 / 1971 / 1972 / 1973 / 1974 / 1975 / 1976 / 1977 / 1978 / 1979 / 1980 / 1981 / 1982 / 1983 / 1984 / 1985 / 1986 / 1987 / 1988 / 1989 / 1990 / 1991 / 1992 / 1993 / 1994 / 1995 / 1996 / 1997 / 1998 / 1999 / 2000 / 2001 / 2002 / 2003 / 2004 / 2005 / 2006 / 2007 / 2008 / 2009 / 2010 / 2011 / 2012 / 2013 / 2014 / 2015 / 2016 / 2017 / 2018 / 2019 / 2020 / 2021 / 2022 / <NA>


In [ ]:
print(*set(df["fuel"]), sep=" / ")

nan / electric / diesel / other / hybrid / gas


In [ ]:
print(*set(df["transmission"]), sep=" / ")

manual / other / nan / automatic


In [ ]:
print(*set(df["paint_color"]), sep=" / ")

yellow / nan / white / black / silver / green / grey / blue / orange / purple / custom / brown / red


In [ ]:
print(*set(df["type"]), sep=" / ")

van / nan / sedan / convertible / wagon / pickup / hatchback / other / coupe / offroad / truck / mini-van / bus / SUV


In [ ]:
print(*set(df["cylinders"]), sep=" / ") 

6 cylinders / 5 cylinders / nan / 4 cylinders / 10 cylinders / other / 8 cylinders / 12 cylinders / 3 cylinders


> OK, data seems fine, we can probably start visualization

# Last - Potential Questions for Data Visualization

**Basic Questions**
- What are the top five manufacturers?
- What are the top 10 models?
- What are the most common vehicle types (SUV, sedan, truck, etc.)?

- What is the distribution of vehicle prices?
- What are the mean, median, minimum, and maximum prices of the vehicles?
- Mean price for top1 vehicle

- What is the distribution of vehicle manufacturing years?
- What are the newest and oldest models in the dataset?

- What is the distribution of the most common energy types (gasoline/diesel vs. electric vs. other)?
- What are the most common transmission types (automatic, manual, etc.)?
- What are the most common paint colors?

**Correlation Questions**
- What is the correlation between vehicle prices and other variables such as manufacturing year, mileage, and engine size (cylinders)? -> _Heatmap!!_

**NLP Exploration**
- What are the most frequent keywords in the vehicle descriptions?
- Are there specific keywords associated with higher or lower prices?
- What is the general sentiment of the vehicle descriptions (positive, negative, neutral)?
- Are there specific topics that are more frequent for certain vehicle types or manufacturers?